### PR TITLE
Fga 3 fga 4

### PR DESCRIPTION
**Pull request by MasterKiller1239**
-User class added
-UserCard class added
-UsersList class added
-Search mechanism added

### CODE REVIEW

File: gitapp/lib/users_list/users_list.dart [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+          avatarUrl: 'https://i.kym-cdn.com/entries/icons/original/000/035/310/Peepo_Animation_Banner.jpg',
+          username: "User"+i.toString());
+      userlist.add(user);
+    }
+
+  }
+
+void createUsers()  {
+  userlist.clear();
+for(int i=0;i<5;i++)
```
Comment (krecikzgr) [inline gitapp/lib/users_list/users_list.dart]: Try to use `map` function instead of for.

File: gitapp/lib/main.dart [hunk 2]
```diff
@@ -10,20 +12,14 @@ class MyApp extends StatelessWidget {
...
-        // application has a blue toolbar. Then, without quitting the app, try
-        // changing the primarySwatch below to Colors.green and then invoke
-        // "hot reload" (press "r" in the console where you ran "flutter run",
-        // or simply save your changes to "hot reload" in a Flutter IDE).
-        // Notice that the counter didn't reset back to zero; the application
-        // is not restarted.
-        primarySwatch: Colors.blue,
+      title: 'Gitapp',
+      theme: ThemeData.dark().copyWith(
+        primaryColor: Color(0xFF2D239F),
```
Comment (krecikzgr) [inline gitapp/lib/main.dart]: If we have added a `constants` file with colors, then please use the colors from that file instead  color in here.

File: gitapp/lib/main.dart [hunk 3]
```diff
@@ -46,19 +42,82 @@ class MyHomePage extends StatefulWidget {
   _MyHomePageState createState() => _MyHomePageState();
 }
 
-class _MyHomePageState extends State<MyHomePage> {
-  int _counter = 0;
+class _MyHomePageState extends State<MyHomePage> with TickerProviderStateMixin {
+  //FGA-3
```
Comment (krecikzgr) [inline gitapp/lib/main.dart]: Comment not needed

File: gitapp/lib/main.dart [hunk 4]
```diff
@@ -10,20 +12,14 @@ class MyApp extends StatelessWidget {
...
-        primarySwatch: Colors.blue,
+      title: 'Gitapp',
+      theme: ThemeData.dark().copyWith(
+        primaryColor: Color(0xFF2D239F),
+        scaffoldBackgroundColor: Color(0xFF313EBB),
+
+        brightness: Brightness.dark
       ),
-      home: MyHomePage(title: 'Flutter Demo Home Page'),
+      home: MyHomePage(title: 'Gitapp'),
```
Comment (krecikzgr) [inline gitapp/lib/main.dart]: please extract `MyHomePage` to another file. Also we should not use anything like `My`. My means whose?
It is not mine, yours. It is app state so it should be called simple `HomePage`

File: gitapp/lib/main.dart [hunk 5]
```diff
@@ -46,19 +42,82 @@ class MyHomePage extends StatefulWidget {
...
     setState(() {
-      // This call to setState tells the Flutter framework that something has
-      // changed in this State, which causes it to rerun the build method below
-      // so that the display can reflect the updated values. If we changed
-      // _counter without calling setState(), then the build method would not be
-      // called again, and so nothing would appear to happen.
-      _counter++;
+      searching = true;
     });
+   if(text!="")
```
Comment (krecikzgr) [inline gitapp/lib/main.dart]: Please fix the formatting.
Prefer `if () {` so the braces should be in the same line

File: gitapp/lib/main.dart [hunk 6]
```diff
@@ -46,19 +42,82 @@ class MyHomePage extends StatefulWidget {
...
+     {
+       name= text;
+       _textController.clear();
+       users.GetSearchedUsers(text);
+       if(users.searchedlist !=null)
+       {
+
+         for (User user in users.searchedlist)
```
Comment (krecikzgr) [inline gitapp/lib/main.dart]: instead of `for` method we could use `map` function, please do so

File: gitapp/lib/main.dart [hunk 7]
```diff
@@ -46,19 +42,82 @@ class MyHomePage extends StatefulWidget {
...
+         {
+           UserCard card = new UserCard(
+             user:user.username,
+             image: user.avatarUrl,
+
+
+             animationController: new AnimationController(
+               duration: new Duration(milliseconds: 700),
```
Comment (krecikzgr) [inline gitapp/lib/main.dart]: If we are using constants then please also move those milliseconds to contants

File: gitapp/lib/main.dart [hunk 8]
```diff
@@ -46,19 +42,82 @@ class MyHomePage extends StatefulWidget {
...
+     }
+  }
+  void dispose() {
+    for (UserCard message in _card)
+      message.animationController.dispose();
+    super.dispose();
   }
+    Widget _buildTextComposer() {
+      return new IconTheme(
+          data: new IconThemeData(color: Theme.of(context).accentColor),
```
Comment (krecikzgr) [inline gitapp/lib/main.dart]: Please do not use `new` keyword. Probably it is not needed.

File: gitapp/lib/main.dart [hunk 9]
```diff
@@ -46,19 +42,82 @@ class MyHomePage extends StatefulWidget {
...
+          child: new Container(
+            margin: const EdgeInsets.symmetric(horizontal: 8.0),
+            child: new Row(children: <Widget>[
+              //FGA-4
```
Comment (krecikzgr) [inline gitapp/lib/main.dart]: please remove comment

File: gitapp/lib/main.dart [hunk 10]
```diff
@@ -46,19 +42,82 @@ class MyHomePage extends StatefulWidget {
...
+              new Flexible(
+                child: new TextField(
+                  controller: _textController,
+                  onSubmitted: _getUsers,
+                  cursorColor: Colors.blueGrey,
```
Comment (krecikzgr) [inline gitapp/lib/main.dart]: please extract `_getUsers` to some external class, according to `MVP`pattern.

File: gitapp/lib/main.dart [hunk 11]
```diff
@@ -74,36 +133,25 @@ class _MyHomePageState extends State<MyHomePage> {
...
+                new Container(
+                  decoration: new BoxDecoration(
+                      color: Theme.of(context).cardColor),
+                  child: _buildTextComposer(),
+                ),
+                new Divider(height: 2.0),
+                new Flexible(
+                    child: new ListView.builder(
+                      padding: new EdgeInsets.all(8.0),
+                      itemBuilder: (_, int index) => _card[index],
```
Comment (krecikzgr) [inline gitapp/lib/main.dart]: `_card` suggests that this is one item. In reality it is more than one, so use plural. Remember to name the variables properly.

File: gitapp/lib/user_details/user_details.dart [hunk 12]
```diff
@@ -13,9 +13,73 @@ class DetailScreen extends StatelessWidget {
...
-        padding: EdgeInsets.all(16.0),
-        child: Text("User details screen"),
+      body: Container(
+        padding: EdgeInsets.all(30.0),
+        child: Row(
+            crossAxisAlignment: CrossAxisAlignment.start,
+            mainAxisAlignment: MainAxisAlignment.start,
+            children: <Widget> [
+              Container(
+
```
Comment (krecikzgr) [inline gitapp/lib/user_details/user_details.dart:23]: formatting

File: gitapp/lib/user_details/user_details.dart [hunk 13]
```diff
@@ -13,9 +13,73 @@ class DetailScreen extends StatelessWidget {
...
+                decoration: BoxDecoration(
+                  shape: BoxShape.circle,
+                  border: Border.all(width: 3.0, color: Colors.white),
+                  image: DecorationImage(
+                    fit: BoxFit.fill,
+                    image: NetworkImage('https://i.kym-cdn.com/entries/icons/original/000/035/310/Peepo_Animation_Banner.jpg'),
+                  ),
+                ),
+              ),
+              // Center(
```
Comment (krecikzgr) [inline gitapp/lib/user_details/user_details.dart:35]: please remove the comments

File: gitapp/lib/user_details/user_details.dart [hunk 14]
```diff
@@ -13,9 +13,73 @@ class DetailScreen extends StatelessWidget {
...
+            crossAxisAlignment: CrossAxisAlignment.start,
+            mainAxisAlignment: MainAxisAlignment.start,
+            children: <Widget> [
+              Container(
+
+                height: 150.0,
+                width: 150.0,
+                decoration: BoxDecoration(
+                  shape: BoxShape.circle,
+                  border: Border.all(width: 3.0, color: Colors.white),
```
Comment (krecikzgr) [inline gitapp/lib/user_details/user_details.dart:28]: if you are using white colour, then still please move it to constants. In this way we will have all the colours in one place and under control.

File: gitapp/lib/user_details/user_details.dart [hunk 15]
```diff
@@ -13,9 +13,73 @@ class DetailScreen extends StatelessWidget {
       appBar: AppBar(
         title: Text("User details"),
       ),
-      body: Padding(
-        padding: EdgeInsets.all(16.0),
-        child: Text("User details screen"),
+      body: Container(
```
Comment (krecikzgr) [inline gitapp/lib/user_details/user_details.dart:16]: Maybe we could extract this widget to single class ??

File: gitapp/lib/users_list/User_card.dart [hunk 16]
```diff
@@ -0,0 +1,69 @@
+import 'package:flutter/material.dart';
+import 'package:gitapp/constants/constants.dart';
+import 'package:gitapp/user_details/user_details.dart';
+
+//card of the user FGA-3
```
Comment (krecikzgr) [inline gitapp/lib/users_list/User_card.dart]: please remove comment

File: gitapp/lib/users_list/user.dart [hunk 17]
```diff
@@ -0,0 +1,9 @@
+//class of the user FGA-3
+class User {
+  String username;
+  String avatarUrl;
+
+  User({required this.avatarUrl, required this.username});
```
Comment (krecikzgr) [inline gitapp/lib/users_list/user.dart]: Formatting

File: gitapp/lib/users_list/users_list.dart [hunk 18]
```diff
@@ -0,0 +1,45 @@
+import 'package:gitapp/users_list/user.dart';
+//Lists of the users FGA-3
```
Comment (krecikzgr) [inline gitapp/lib/users_list/users_list.dart]: please remove the comment

File: gitapp/lib/users_list/users_list.dart [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+class UsersList {
+  List<User> userlist= List.empty(growable: true);
+  List<User> searchedlist= List.empty(growable: true);
+  UsersList(){
+    for(int i=0;i<5;i++)
```
Comment (krecikzgr) [inline gitapp/lib/users_list/users_list.dart]: please do not use this for method.
Use map, foreach or something similar. Let's work with higher order functions directly on collections.

File: gitapp/lib/users_list/users_list.dart [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+    }
+
+  }
+
+void createUsers()  {
+  userlist.clear();
+for(int i=0;i<5;i++)
+  {
+
+    User user = new User(
```
Comment (krecikzgr) [inline gitapp/lib/users_list/users_list.dart]: Do we have some duplicated functionality ?
the `createUsers` and `UsersList` ?

File: gitapp/lib/users_list/users_list.dart [hunk 21]
```diff
@@ -0,0 +1,45 @@
...
+        avatarUrl: 'https://i.kym-cdn.com/entries/icons/original/000/035/310/Peepo_Animation_Banner.jpg',
+        username: "User"+i.toString());
+    userlist.add(user);
+  }
+
+
+}
+
+  void GetSearchedUsers(String text)
```
Comment (krecikzgr) [inline gitapp/lib/users_list/users_list.dart]: name of the method has to start from lowercased letter.
Also if you are saying `get` then it suggests that the method should return something, but in fact it does not.

File: gitapp/lib/HomePage/homepage.dart [hunk 22]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+class HomePageState extends State<HomePage> with TickerProviderStateMixin {
+  List<UserCard> _card = <UserCard>[];
+  final UsersList users = new UsersList();
+  String name = "";
+  final TextEditingController _textController = new TextEditingController();
+  bool searching = false;
+
+  Future _getUsers(String text) async {
```
Comment (krecikzgr) [inline gitapp/lib/HomePage/homepage.dart:22]: Have we already migrated to Presenter?

File: gitapp/lib/HomePage/homepage.dart [hunk 23]
```diff
@@ -0,0 +1,119 @@
...
+      name = text;
+      _textController.clear();
+
+      //users.fillSearchedUsers(text);
+      await users.fetchUsers(text);
+
+      if (users.searchedlist.length != 0) {
+        users.searchedlist.forEach((element) {
+          setState(() {
+            _card.insert(
```
Comment (krecikzgr) [inline gitapp/lib/HomePage/homepage.dart]: There should not be a situation like this.
We can't create a widget for every user which has been found.
Consider that kind of situation. My search result is quite big and I've found 1000 elements. In here you are iterating over this whole collection and you are creating an widget for all of elements. And you are loading 1000 elements into memory. Usage of ListViewBuilder allows us to use much smarter mechanism. ListViewBuilder itself asks for now element every time the users scrolls on it. And in that listviewbuilder we should return a widget only when user will scroll. Thank's to that our app will be much faster. Please refer to documentation, ask Bartek about details or wait for me for tommorow,  we can make a call and I will explain the problem in here.

File: gitapp/lib/HomePage/homepage.dart [hunk 24]
```diff
@@ -0,0 +1,119 @@
...
+
+    if (text != "") {
+      name = text;
+      _textController.clear();
+
+      //users.fillSearchedUsers(text);
+      await users.fetchUsers(text);
+
+      if (users.searchedlist.length != 0) {
+        users.searchedlist.forEach((element) {
```
Comment (krecikzgr) [inline gitapp/lib/HomePage/homepage.dart]: Also we should not use foreach.
If We want to transform an collection of items into a collection of items of different type, we should use map. method. 

Please refer to the following link.

https://medium.com/flutter-community/useful-list-methods-in-dart-6e173cac803d

File: gitapp/lib/HomePage/homepage.dart [hunk 25]
```diff
@@ -0,0 +1,119 @@
...
+  }
+
+  void dispose() {
+    for (UserCard message in _card) message.animationController.dispose();
+    super.dispose();
+  }
+
+  Widget _buildTextComposer() {
+    return new IconTheme(
+        data: new IconThemeData(color: Theme.of(context).accentColor),
```
Comment (krecikzgr) [inline gitapp/lib/HomePage/homepage.dart]: Please remove all "new" keyword occuernces  from the project

File: gitapp/lib/constants/constants.dart [hunk 26]
```diff
@@ -0,0 +1,8 @@
+import 'package:flutter/material.dart';
+
+const primaryLight = Color(0xFF313EBB);
+const primaryDark = Color(0xFF2D239F);
+const secondaryLight = Colors.blueAccent;
+const animationtime = 700;
+const cursorColor = Colors.blueGrey;
+const NotProvided = 'Not provided 😔';
```
Comment (krecikzgr) [inline gitapp/lib/constants/constants.dart]: Please use one common approach of naming. Lowercase, or uppercase, do not mix it.

File: gitapp/lib/main.dart [hunk 27]
```diff
@@ -1,5 +1,8 @@
 import 'package:flutter/material.dart';
-import 'package:gitapp/user_details/user_details.dart';
+
+
```
Comment (krecikzgr) [inline gitapp/lib/main.dart]: Are we sure that formatting is fixed?

File: gitapp/lib/users_list/User_card.dart [hunk 28]
```diff
@@ -0,0 +1,65 @@
+import 'package:flutter/material.dart';
+import 'package:gitapp/constants/constants.dart';
+import 'package:gitapp/user_details/user_details.dart';
+
+class UserCard extends StatelessWidget {
+  UserCard(
+      {required this.user,
```
Comment (krecikzgr) [inline gitapp/lib/users_list/User_card.dart]: I think that we could pass only an User object which could contain the UserName, and avatar image.

File: gitapp/lib/users_list/User_card.dart [hunk 29]
```diff
@@ -0,0 +1,65 @@
...
+      required this.image,
+      required this.animationController});
+
+  final String user;
+  final String image;
+
+  final AnimationController animationController;
+
+  @override
```
Comment (krecikzgr) [inline gitapp/lib/users_list/User_card.dart]: Formatting

File: gitapp/lib/users_list/User_card.dart [hunk 30]
```diff
@@ -0,0 +1,65 @@
...
+                    MaterialPageRoute(builder: (context) => DetailScreen()));
+              },
+              child: Container(
+                padding: EdgeInsets.symmetric(vertical: 12.0),
+                margin: const EdgeInsets.only(right: 5.0),
+                child: Row(
+                  crossAxisAlignment: CrossAxisAlignment.start,
+                  children: <Widget>[
+                    Container(
+                      margin: const EdgeInsets.only(right: 16.0),
```
Comment (krecikzgr) [inline gitapp/lib/users_list/User_card.dart]: Do we need const keyword?

File: gitapp/lib/users_list/users_list.dart [hunk 31]
```diff
@@ -0,0 +1,46 @@
+import 'package:flutter/material.dart';
+import 'package:gitapp/users_list/user.dart';
+import "package:http/http.dart" as http;
+import 'dart:convert' as convert;
+
+class UsersList {
+  List<User> userlist = List.empty(growable: true);
+  List<User> searchedlist = List.empty(growable: true);
```
Comment (krecikzgr) [inline gitapp/lib/users_list/users_list.dart]: Please use camel case in all examples so 

searchedList

File: gitapp/lib/users_list/users_list.dart [hunk 32]
```diff
@@ -0,0 +1,46 @@
...
+
+  UsersList() {
+    List<String> usersnames = ['Tytus', 'Romek', 'Atomek', 'Bolek', 'Lolek'];
+    userlist = usersnames
```
Comment (krecikzgr) [inline gitapp/lib/users_list/users_list.dart]: as above

File: gitapp/lib/users_list/users_list.dart [hunk 33]
```diff
@@ -0,0 +1,46 @@
...
+        .map((name) => User(
+            username: name,
+            avatarUrl:
+                'https://i.kym-cdn.com/entries/icons/original/000/035/310/Peepo_Animation_Banner.jpg'))
```
Comment (krecikzgr) [inline gitapp/lib/users_list/users_list.dart]: I think that in github API we should have the avatar URL

File: gitapp/lib/users_list/users_list.dart [hunk 34]
```diff
@@ -0,0 +1,46 @@
...
+
+  Future<bool> fetchUsers(String username) async {
+    String httpadress =
+        'https://api.github.com/search/users?q=$username&per_page=5';
+
+    searchedlist.clear();
+    http.Response response = await http.get(Uri.parse(httpadress));
+    if (response.statusCode == 200) {
+      var data = convert.jsonDecode(response.body);
+
```
Comment (krecikzgr) [inline gitapp/lib/users_list/users_list.dart]: Formatting

File: gitapp/lib/users_list/users_list.dart [hunk 35]
```diff
@@ -0,0 +1,46 @@
...
+      data['items'].forEach((user) {
+        searchedlist.add(User.fromJSON(user));
+      });
+      return true;
+    }
+    return false;
+  }
+
+  void fillSearchedUsers(String text) {
```
Comment (krecikzgr) [inline gitapp/lib/users_list/users_list.dart]: what does fillSearchedUsers do ?

File: gitapp/lib/main.dart (hunk 2 above)
Comment (MasterKiller1239) [reply gitapp/lib/main.dart]: solved

File: gitapp/lib/users_list/users_list.dart (hunk 1 above)
Comment (MasterKiller1239) [reply gitapp/lib/users_list/users_list.dart]: solved

File: gitapp/lib/main.dart (hunk 3 above)
Comment (MasterKiller1239) [reply gitapp/lib/main.dart]: solved

File: gitapp/lib/main.dart (hunk 4 above)
Comment (MasterKiller1239) [reply gitapp/lib/main.dart]: solved

File: gitapp/lib/main.dart (hunk 5 above)
Comment (MasterKiller1239) [reply gitapp/lib/main.dart]: solved

File: gitapp/lib/main.dart (hunk 7 above)
Comment (MasterKiller1239) [reply gitapp/lib/main.dart]: solved

File: gitapp/lib/main.dart (hunk 8 above)
Comment (MasterKiller1239) [reply gitapp/lib/main.dart]: solved

File: gitapp/lib/main.dart (hunk 9 above)
Comment (MasterKiller1239) [reply gitapp/lib/main.dart]: solved

File: gitapp/lib/user_details/user_details.dart (hunk 12 above)
Comment (MasterKiller1239) [reply gitapp/lib/user_details/user_details.dart:23]: not mine

File: gitapp/lib/users_list/User_card.dart (hunk 16 above)
Comment (MasterKiller1239) [reply gitapp/lib/users_list/User_card.dart]: solved

File: gitapp/lib/users_list/user.dart (hunk 17 above)
Comment (MasterKiller1239) [reply gitapp/lib/users_list/user.dart]: solved

File: gitapp/lib/users_list/users_list.dart (hunk 18 above)
Comment (MasterKiller1239) [reply gitapp/lib/users_list/users_list.dart]: solved

File: gitapp/lib/users_list/users_list.dart (hunk 19 above)
Comment (MasterKiller1239) [reply gitapp/lib/users_list/users_list.dart]: solved

File: gitapp/lib/users_list/users_list.dart (hunk 20 above)
Comment (MasterKiller1239) [reply gitapp/lib/users_list/users_list.dart]: solved

File: gitapp/lib/users_list/users_list.dart (hunk 21 above)
Comment (MasterKiller1239) [reply gitapp/lib/users_list/users_list.dart]: solved

File: gitapp/lib/HomePage/homepage.dart (hunk 25 above)
Comment (MasterKiller1239) [reply gitapp/lib/HomePage/homepage.dart]: solved

File: gitapp/lib/constants/constants.dart (hunk 26 above)
Comment (MasterKiller1239) [reply gitapp/lib/constants/constants.dart]: solved

File: gitapp/lib/main.dart (hunk 27 above)
Comment (MasterKiller1239) [reply gitapp/lib/main.dart]: solved

File: gitapp/lib/users_list/User_card.dart (hunk 29 above)
Comment (MasterKiller1239) [reply gitapp/lib/users_list/User_card.dart]: solved

File: gitapp/lib/users_list/User_card.dart (hunk 30 above)
Comment (MasterKiller1239) [reply gitapp/lib/users_list/User_card.dart]: solved

File: gitapp/lib/users_list/users_list.dart (hunk 31 above)
Comment (MasterKiller1239) [reply gitapp/lib/users_list/users_list.dart]: solved

File: gitapp/lib/users_list/users_list.dart (hunk 32 above)
Comment (MasterKiller1239) [reply gitapp/lib/users_list/users_list.dart]: solved

File: gitapp/lib/users_list/users_list.dart (hunk 34 above)
Comment (MasterKiller1239) [reply gitapp/lib/users_list/users_list.dart]: solved

File: gitapp/lib/users_list/users_list.dart (hunk 35 above)
Comment (MasterKiller1239) [reply gitapp/lib/users_list/users_list.dart]: solved

File: gitapp/lib/main.dart (hunk 6 above)
Comment (MasterKiller1239) [reply gitapp/lib/main.dart]: solved

File: gitapp/lib/main.dart (hunk 10 above)
Comment (MasterKiller1239) [reply gitapp/lib/main.dart]: Solved

File: gitapp/lib/main.dart (hunk 11 above)
Comment (MasterKiller1239) [reply gitapp/lib/main.dart]: Solved

File: gitapp/lib/user_details/user_details.dart (hunk 13 above)
Comment (MasterKiller1239) [reply gitapp/lib/user_details/user_details.dart:35]: not mine

File: gitapp/lib/user_details/user_details.dart (hunk 14 above)
Comment (MasterKiller1239) [reply gitapp/lib/user_details/user_details.dart:28]: Solved

File: gitapp/lib/user_details/user_details.dart (hunk 15 above)
Comment (MasterKiller1239) [reply gitapp/lib/user_details/user_details.dart:16]: not mine

File: gitapp/lib/users_list/User_card.dart (hunk 28 above)
Comment (MasterKiller1239) [reply gitapp/lib/users_list/User_card.dart]: Solved

File: gitapp/lib/users_list/users_list.dart (hunk 33 above)
Comment (MasterKiller1239) [reply gitapp/lib/users_list/users_list.dart]: Solved

File: gitapp/lib/HomePage/homepage.dart (hunk 24 above)
Comment (MasterKiller1239) [reply gitapp/lib/HomePage/homepage.dart]: Solved

File: gitapp/lib/HomePage/homepage.dart (hunk 23 above)
Comment (MasterKiller1239) [reply gitapp/lib/HomePage/homepage.dart]: Solved

File: gitapp/lib/HomePage/homepage.dart (hunk 22 above)
Comment (MasterKiller1239) [reply gitapp/lib/HomePage/homepage.dart:22]: I guess solved

**Pull request merged**